packages/react-select/src/internal/ScrollManager.js

```diff
@@ -0,0 +1,189 @@
+    }
+    if (typeof el.addEventListener === 'function') {
+      el.addEventListener('touchmove', this.onTouchMove, false);
+    }
+  }
+
+  stopListening(el: ?HTMLElement) {
+    // bail early if no element is available to detach from
+    if (!el) return;
+
+    // all the if statements are to appease Flow 😢
+    if (typeof el.removeEventListener === 'function') {
+      el.removeEventListener('wheel', this.onWheel, false);
+    }
+    if (typeof el.removeEventListener === 'function') {
+      el.removeEventListener('touchstart', this.onTouchStart, false);
+    }
+    if (typeof el.removeEventListener === 'function') {
+      el.removeEventListener('touchmove', this.onTouchMove, false);
+    }
+  }
+
+  cancelScroll = (event: SyntheticEvent<HTMLElement>) => {
+    event.preventDefault();
+    event.stopPropagation();
+  };
+  handleEventDelta = (event: SyntheticEvent<HTMLElement>, delta: number) => {
+    const {
+      onBottomArrive,
+      onBottomLeave,
+      onTopArrive,
+      onTopLeave,
+    } = this.props;
+
+    // Reference should never be `null` at this point, but flow complains otherwise
+    if (this.targetRef.current === null) return;
+
+    const { scrollTop, scrollHeight, clientHeight } = this.targetRef.current;
+    const target = this.targetRef.current;
+    const isDeltaPositive = delta > 0;
+    const availableScroll = scrollHeight - clientHeight - scrollTop;
+    let shouldCancelScroll = false;
+
+    // reset bottom/top flags
+    if (availableScroll > delta && this.isBottom) {
+      if (onBottomLeave) onBottomLeave(event);
+      this.isBottom = false;
+    }
+    if (isDeltaPositive && this.isTop) {
+      if (onTopLeave) onTopLeave(event);
+      this.isTop = false;
+    }
+
+    // bottom limit
+    if (isDeltaPositive && delta > availableScroll) {
+      if (onBottomArrive && !this.isBottom) {
+        onBottomArrive(event);
+      }
+      target.scrollTop = scrollHeight;
+      shouldCancelScroll = true;
+      this.isBottom = true;
+
+      // top limit
+    } else if (!isDeltaPositive && -delta > scrollTop) {
+      if (onTopArrive && !this.isTop) {
+        onTopArrive(event);
+      }
+      target.scrollTop = 0;
+      shouldCancelScroll = true;
+      this.isTop = true;
+    }
+
+    // cancel scroll
+    if (shouldCancelScroll) {
+      this.cancelScroll(event);
+    }
+  };
+
+  onWheel = (event: SyntheticWheelEvent<HTMLElement>) => {
+    this.handleEventDelta(event, event.deltaY);
+  };
+  onTouchStart = (event: SyntheticTouchEvent<HTMLElement>) => {
+    // set touch start so we can calculate touchmove delta
+    this.touchStart = event.changedTouches[0].clientY;
+  };
+  onTouchMove = (event: SyntheticTouchEvent<HTMLElement>) => {
+    const deltaY = this.touchStart - event.changedTouches[0].clientY;
+    this.handleEventDelta(event, deltaY);
+  };
+
+  setTargetRef = (instance: HTMLElement | null) => {
+    this.targetRef.current = instance;
+    this.setState({ enableLock: !!instance });
```

Comment:
> If the intent is to trigger a re-render, can we store the element in state, similar to what ScrollBlock does?

>  I know you've said you're against storing an element in state, I'd be curious to know why.

I learned and made it my ideology, that `state` is **explicitly** for data, while references to DOM elements and possibly React components should be kept in `refs`. That is why I`m against it. Period.

> Another option would be to make ScrollLock not a React component anymore since all it does is attach event listeners to the element and it doesn't actually render anything.

The React mindset is all about components. And I think either would not be an improvement over the other.

Reply:
That's fine, I would prefer not putting DOM elements in `state` as well (although I think that's sometimes the best solution available, just not in this case).

While the React mindset is all about components, that doesn't mean literally every piece of logic has to live in a component. The design of `react-select` is clearly not against taking logic that makes most sense in a `componentDidMount`, etc. and abstracting it into a function based on `utils.js`.

I would also argue that the mindset of React is also against taking DOM elements received from refs and passing them down to child components through props, because it makes it difficult to make sure that those elements are properly received without forcing what should be an unnecessary re-render on the parent component. The reason why `componentDidMount`, etc. exist are to perform actions on DOM elements that are received via refs after the `render` has taken place, but that breaks down as soon as you start passing elements received from refs as props to other components.

Reply:
Also, components that return `null` (like `ScrollLock`) don't really fit into the React mindset either.

Reply:
Back to topic: I don't think it's worth the effort to rewrite `ScrollLock` (for this PR). Maybe add it to the backlog?

Reply:
Maybe we could rewrite `ScrollLock` to be a wrapping component that also has its own reference to the DOM element.

Or we could use a library for that: https://github.com/jossmac/react-scrolllock

Reply:
Sounds fine, I'll make a follow-up PR to avoid the re-render in `ScrollManager` and hopefully we can get them merged in back-to-back for 4.0, or if there is a problem with the follow-up PR, we can just wait to merge it in order to get this released ASAP.

---

Would we have to use the same pattern of having a `(ref) => { }` child function if it were a wrapping component? That seems like a necessary pattern in these use cases, but I personally think it adds unnecessary complexity when you're doing it more than once for the same DOM element. I understand the intention to abstract the functionality it into a reusable component, but I think there's a more simple way to achieve this behavior that follows good coding practices and is still in the React mindset (imo).

Yeah, using a library sounds good too, although I'm not sure I understand how it works if the child is a component and not a function because it seems like it's just getting a ref to the child component and not a ref to the DOM element (based on [this line](https://github.com/jossmac/react-scrolllock/blob/master/src/TouchScrollable.js#L52))?